gax/pagination.h

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_PAGINATION_H_
+#define GAPIC_GENERATOR_CPP_GAX_PAGINATION_H_
+
+#include <google/protobuf/repeated_field.h>
+#include <functional>
+#include <iterator>
+#include <string>
+
+namespace google {
+namespace gax {
+
+template <typename ElementType, typename PageType>
+class PageResult {
+  using ElementAccessor =
+      std::function<::google::protobuf::RepeatedPtrField<ElementType>*(
+          PageType&)>;
+
+ public:
+  class iterator {
+   public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = ElementType;
+    using difference_type = std::ptrdiff_t;
+    using pointer = ElementType*;
+    using reference = ElementType&;
+
+    reference operator*() const { return *current_; }
+    pointer operator->() const { return &(*current_); }
+    iterator& operator++() {
+      ++current_;
+      return *this;
+    }
+    bool operator==(iterator const& rhs) const {
+      return current_ == rhs.current_;
+    }
+    bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
+
+    iterator(typename ::google::protobuf::RepeatedPtrField<
+             ElementType>::iterator current)
+        : current_(std::move(current)) {}
+
+    typename ::google::protobuf::RepeatedPtrField<ElementType>::iterator
+        current_;
+  };
+  // TODO: Need to take into account RepeatedField vs RepeatedPtrField.
+  // They have different iterator types,
+  // and will probably need some templating magic.
+  // Note: Since the constructor will eventually be private (with friend
+  // access), we don't have to define all the variants for references.
+  // We just want to take ownership of a page.
+  PageResult(PageType&& page, ElementAccessor accessor)
```

Comment:
`accessor` is used solely to call  `->begin()` and `->end()` on page within this constructor implementation (i.e. accessor itself is not stored as a fiedl inside `PageResult` class). Looks like a convoluted way of doing a trivial thing (i.e. complications for the users of this class without benefits).  Maybe just ask for begin and end iterators here directly?

Reply:
This section is being rewritten to store the accessor as an attribute that will (hopefully) be inlined and elided. The begin() and end() iterators will then be generated dynamically instead of copied.